### PR TITLE
Extra Reset necessary when looping EventFiles

### PR DESCRIPTION
**Pull request by tomeichlersmith**
I am updating _ldmx-sw_, here are the details.

### What are the issues that this addresses?
This popped up when working with @bryngemark to debug some seg-faulting errors they were seeing while beta-testing the overlay code with the (almost done) v3.0.0 release.

## Check List
- [x] I successfully compiled _ldmx-sw_ with my developments
- [x] I ran my developments and the following shows that they are successful.
I was able to run the attached configuration script `config.py` using the following two files for the overlay.
```
/nfs/slac/g/ldmx/data/mc20/v12/4.0GeV/v2.3.0-batch1/mc_v12-4GeV-1e-ecal_photonuclear_run1067_t1608081969.root # main event
/nfs/slac/g/ldmx/data/mc20/v12/4.0GeV/v2.3.0-1e/mc_v12-4GeV-1e-inclusive_run10076_t1612472444.root # pileup
```

- [x] I attached any sub-module related changes to this PR.

### Related Sub-Module PRs
- https://github.com/LDMX-Software/Framework/pull/45

## config.py
```python
#!/usr/bin/python
​
import sys
import os
​
from LDMX.Framework import ldmxcfg
​
# first, we define the process, which must have a name which identifies this
# processing pass ("pass name").
simPassName=""
pileupFilePassName=""
thisPassName="overlay"
p=ldmxcfg.Process(thisPassName)
​
from LDMX.Recon.overlay import OverlayProducer
overlay=OverlayProducer('pileup.root')
overlay.passName = simPassName                  #sim input event pass name
overlay.overlayPassName = pileupFilePassName    #pileup input event pass name
overlay.totalNumberOfInteractions = 2.
overlay.doPoissonIntime = False
overlay.doPoissonOutoftime = False
overlay.nEarlierBunchesToSample = 0
overlay.bunchSpacing = 26.9      #ns = 1000./37.2 ; 5.4 = 1000./186.
overlay.timeSpread = 0.       # <-- realistically, 30 ps; 
overlay.timeMean   = 0.       # <-- here set the in-bunch average pu time offset to no time shift whatsoever; useful for validation though
overlay.overlayCaloHitCollections=[ "TriggerPadTaggerSimHits", "TriggerPadUpSimHits", "TriggerPadDownSimHits", "TargetSimHits", "EcalSimHits", "HcalSimHits"]
overlay.overlayTrackerHitCollections=[ "TaggerSimHits", "RecoilSimHits" ]
overlay.verbosity=0 #1 #3 #
​
p.sequence = [overlay]
#p.termLogLevel = 1
#p.logFrequency = 100
​
# ECal geometry nonsense
from LDMX.Ecal import EcalGeometry
import LDMX.Ecal.ecal_hardcoded_conditions
​
# Hcal hardwired/geometry stuff
from LDMX.Hcal import HcalGeometry
import LDMX.Hcal.hcal_hardcoded_conditions
​
from LDMX.Ecal import digi as eDigi
from LDMX.Ecal import vetos
from LDMX.Hcal import digi as hDigi
from LDMX.TrigScint.trigScint import TrigScintDigiProducer
from LDMX.TrigScint.trigScint import TrigScintClusterProducer
from LDMX.TrigScint.trigScint import trigScintTrack
​
overlayStr="Overlay"
​
ecalDigi   =eDigi.EcalDigiProducer('ecalDigis')
ecalReco   =eDigi.EcalRecProducer('ecalRecon')
ecalVeto   =vetos.EcalVetoProcessor('ecalVetoBDT')
​
ecalDigi.inputCollName  = ecalDigi.inputCollName+overlayStr
ecalReco.simHitCollName = ecalReco.simHitCollName+overlayStr
ecalReco.digiPassName = thisPassName
ecalVeto.rec_pass_name = thisPassName
​
hcalDigi   =hDigi.HcalDigiProducer('hcalDigis')
hcalDigi.inputCollName  = hcalDigi.inputCollName+overlayStr
hcalReco   =hDigi.HcalRecProducer('hcalRecon')
hcalReco.digiPassName = thisPassName
​
p.sequence.extend([
    ecalDigi, ecalReco, ecalVeto,
    hcalDigi, hcalReco])
    
p.inputFiles = [sys.argv[1]]
p.outputFiles= [sys.argv[2]]
```

### CODE REVIEW

**Comment by tomeichlersmith** [comment]
@bryngemark I also made a container with this patch for testing within your batch framework.
`tomeichlersmith/ldmx-sw:loop-reset`
I've downloaded this image to SLAC as well:
`/nfs/slac/g/ldmx/users/eichl008/tomeichlersmith_ldmx-sw_loop-reset.sif`

**Comment by tomeichlersmith** [comment]
Ok, this is truly funky. I'm still seeing seg-faults when I run the production image I mention above, but when I run the _exact_ same branch but compile it myself using the dev container, I don't get a seg-fault. The only difference between these two containers that I can see is a change in the ROOT version (6.22/03 -> 6.22/07). I am building another production image for this branch using the updated development container (v2.0 or latest) which I was using when I wasn't getting a seg-fault.

**Comment by tomeichlersmith** [comment]
A quote from ROOT's [6.22/04 release notes](https://root.cern/doc/v622/release-notes.html#release-6.2204):
> Published on an unlucky Friday, November 13, 2020, this release had a fatal issue in the I/O subsystem, which is why it was never announced. Please use 6.22/06 instead!

This was just bad timing. I have been using a running branch `6-22-00-patches` to get the latest patches on top of 6.22 inside of our container image, so we ended up with 6.22/03 inside the v1.0 dev image which seems to have had this fatal I/O issue. I am working on [developing a v3.0 dev image](https://github.com/LDMX-Software/docker/tree/v3.0-dev) which will cement the versions we are building on so issues like this are less likely to happen in the future.

However, testing this hypothesis by using the `tomeichlersmith/ldmx-sw:loop-reset-on-v2.0` image (which is this branch built on dev v2.0 instead of v1.0) still fails by hitting a seg-fault :disappointed: Going to keep investigating...

**Comment by tomeichlersmith** [comment]
Good news (?)

I got the same seg-fault using docker on my personal laptop (after downloading the files I am testing with from SLAC). This is good news because it rules out the docker->singularity conversion step as a potential source of the error.

Thank you @JHiltbrand for discussion this morning.

I have a fear that ROOT is doing some asynchronous memory-management nonsense in the background and when we are operating very quickly (not many printouts) _and_ try to reset a tree to its beginning, we end up introducing this bug where ROOT's asynchronous memory-management nonsense thinks it needs to clean-up after itself and tries to delete an object that is already deleted. :thinking: :thought_balloon: This makes me think we should reset less things. Playing with ROOT re-enforces this hypothesis:
```cpp
// looping.C
#include <TFile.h>
#include <TTree.h>

void looping() {
  { //write a dummy event file to read
    TFile write_f("dummy.root","recreate");
    TTree write_t("events","looping events");
    int i{0};
    write_t.Branch("index",&i);
    for (;i<5;i++) {
      write_t.Fill();
    }
    write_f.Write();
    write_f.Close();
  }

  { //loop over dummy event file 2.5 times
    TFile read_f("dummy.root");
    auto t{(TTree*)read_f.Get("events")};
    t->SetCacheSize(4); // shrink buffer size to force re-allocation
    int i; // different i because different scope
    t->SetBranchAddress("index",&i);
    for (int j{0}; j < 12; j++) {
      t->GetEntry(j%5);
      std::cout << i << std::endl;
      if (i != j%5) {
        std::cout << "This statement is never printed!" << std::endl;
      }
    }
  }
}
```
Or using a short 10-event file produced by ldmx-sw:
```bash
[ldmx] eichl008@spa-cms015 ~/ldmx> ldmx root
   ------------------------------------------------------------------
  | Welcome to ROOT 6.22/07                        https://root.cern |
  | (c) 1995-2020, The ROOT Team; conception: R. Brun, F. Rademakers |
  | Built for linuxx8664gcc on Feb 28 2021, 18:53:00                 |
  | From heads/v6-22-00-patches@v6-22-06-70-gcbfd8e4405              |
  | Try '.help', '.demo', '.license', '.credits', '.quit'/'.q'       |
   ------------------------------------------------------------------

root [0] gSystem->Load("libFramework.so")
(int) 0
root [1] h = new ldmx::EventHeader;
root [2] TFile f("events.root")
(TFile &) Name: events.root Title: 
root [3] t = (TTree*)f.Get("LDMX_Events")
(TTree *) @0x7ffeeb59b508
root [4] t->SetBranchAddress("EventHeader",&h)
(int) 0
root [5] for (int j{3}; j<12; j++) { t->GetEntry(j%10); h->Print(); }
EventHeader { eventNumber: 4, run: 1, timestamp: 2021-07-06 20:36:45.062703000Z, weight: 1, MC }
EventHeader { eventNumber: 5, run: 1, timestamp: 2021-07-06 20:36:45.311002000Z, weight: 1, MC }
EventHeader { eventNumber: 6, run: 1, timestamp: 2021-07-06 20:36:45.528827000Z, weight: 1, MC }
EventHeader { eventNumber: 7, run: 1, timestamp: 2021-07-06 20:36:45.764231000Z, weight: 1, MC }
EventHeader { eventNumber: 8, run: 1, timestamp: 2021-07-06 20:36:46.002941000Z, weight: 1, MC }
EventHeader { eventNumber: 9, run: 1, timestamp: 2021-07-06 20:36:46.234740000Z, weight: 1, MC }
EventHeader { eventNumber: 10, run: 1, timestamp: 2021-07-06 20:36:46.456132000Z, weight: 1, MC }
EventHeader { eventNumber: 1, run: 1, timestamp: 2021-07-06 20:36:43.562597000Z, weight: 1, MC }
EventHeader { eventNumber: 2, run: 1, timestamp: 2021-07-06 20:36:44.597457000Z, weight: 1, MC }
root [6] .q
```

**Comment by bryngemark** [comment]
So what you're saying here is that we should be able to just loop back to the first event by setting `ientry_ = 0` (or potentially -1, depending on if it gets incremented in later stages)? This is where we started out, I believe. At least this is what I was trying to do 😄  For some reason, the entry actually used was then stuck at the last entry in the tree.

**Comment by tomeichlersmith** [comment]
Yep! I believe that is correct. The newer bus design[0] allows for a more simple interface with ROOT because we handle all of the dynamically-created objects (like I did manually above with the EventHeader) rather than having ROOT handle some of the objects. I think the older design is the reason this more simple looping mechanism failed, but I am still going to test that it works with the files provided and re-run PR validation to make sure the samples get to the plotting stage (the plots will mismatch because we haven't made a new release since merging other plot-changing PRs)

[0] https://github.com/LDMX-Software/Framework/pull/34

File: Recon/src/Recon/OverlayProducer.cxx [hunk 1]
```diff
@@ -213,6 +213,19 @@ void OverlayProducer::produce(framework::Event &event) {
...
+       * return false if an error is occurred or if the overlay file is
+       * mis-configured.
+       */
+      if (!overlayFile_->nextEvent()) {
+        ldmx_log(error) << "At sim event "
+                        << event.getEventHeader().getEventNumber()
+                        << ": couldn't read next overlay event!";
+        return;
+      }
+
```
Comment (bryngemark) [inline Recon/src/Recon/OverlayProducer.cxx:228]: Please double check that the event number pulled is still the same after moving this to the front of processing. I remember moving this down at some point for a reason; likely when I started setting the overlay file up after having initiated the RNSS.

**Comment by tomeichlersmith** [comment]
I just wanted to double check that the overlay producer is looping over the events as expected. I made sure that it ran through the SLAC files given above already.

After compiling this branch, I added a print statement in the OverlayProducer and re-compiled.
```diff
diff --git a/Recon/src/Recon/OverlayProducer.cxx b/Recon/src/Recon/OverlayProducer.cxx
index 40f54f43..bb03698d 100644
--- a/Recon/src/Recon/OverlayProducer.cxx
+++ b/Recon/src/Recon/OverlayProducer.cxx
@@ -226,6 +226,8 @@ void OverlayProducer::produce(framework::Event &event) {
         return;
       }
 
+      overlayEvent_.getEventHeader().Print();
+
       // a pileup event wide time offset to be applied to all its hits.
       float timeOffset = rndmTime_->Gaus(timeMean_, timeSigma_);
       timeOffset += bunchTimeOffset;
```

And then I ran the following command in the `.github/validation_samples/it_pileup` directory.

```bash
ldmx 'export LDMX_NUM_EVENTS=100; 
> export LDMX_RUN_NUMBER=1; 
> for c in gen_main.py gen_pileup.py config.py; do fire $c || break; done'
```

which gave the following output (skipping the most of the un-interesting printout).

```
---- LDMXSW: Starting event processing --------
EventHeader { eventNumber: 60, run: 1, timestamp: 2021-07-07 19:38:26.897504000Z, weight: 1, MC }
EventHeader { eventNumber: 61, run: 1, timestamp: 2021-07-07 19:38:27.400649000Z, weight: 1, MC }
EventHeader { eventNumber: 62, run: 1, timestamp: 2021-07-07 19:38:27.918752000Z, weight: 1, MC }
EventHeader { eventNumber: 63, run: 1, timestamp: 2021-07-07 19:38:28.519341000Z, weight: 1, MC }
EventHeader { eventNumber: 64, run: 1, timestamp: 2021-07-07 19:38:29.028676000Z, weight: 1, MC }
EventHeader { eventNumber: 65, run: 1, timestamp: 2021-07-07 19:38:29.711239000Z, weight: 1, MC }
...keeps counting up in event number...
EventHeader { eventNumber: 93, run: 1, timestamp: 2021-07-07 19:38:45.348903000Z, weight: 1, MC }
EventHeader { eventNumber: 94, run: 1, timestamp: 2021-07-07 19:38:45.971423000Z, weight: 1, MC }
EventHeader { eventNumber: 95, run: 1, timestamp: 2021-07-07 19:38:46.461345000Z, weight: 1, MC }
EventHeader { eventNumber: 1, run: 1, timestamp: 2021-07-07 19:37:52.850851000Z, weight: 1, MC }
EventHeader { eventNumber: 2, run: 1, timestamp: 2021-07-07 19:37:53.670860000Z, weight: 1, MC }
EventHeader { eventNumber: 3, run: 1, timestamp: 2021-07-07 19:37:54.204799000Z, weight: 1, MC }
...keeps counting up in event number...
```
This is the desired behavior: start at non-zero entry index and then wrap back to beginning when we reach the end of the file. (The pileup file in the validation sample is set to have 95% of the events provided by the env variable in order to test this wrapping.)

File: Recon/src/Recon/OverlayProducer.cxx (hunk 1 above)
Comment (bryngemark) [reply Recon/src/Recon/OverlayProducer.cxx:228]: sorry, i see this was never submitted (i forgot to push "submit review"). however, i think this is a minor thing so nothing that would require undoing the merge.

Comment (tomeichlersmith) [reply Recon/src/Recon/OverlayProducer.cxx:228]: Apologies for merging this in before your review! I did double check this and the combination of changing `EventFile::skipToEvent` and moving this to the top of the loop replicated the old process (I just didn't copy down that printout, so you'll have to trust my word :man_shrugging: ).

Comment (bryngemark) [reply Recon/src/Recon/OverlayProducer.cxx:228]: great, thanks!

**Pull request merged**